### PR TITLE
[da-vinci][test] Add a new metric to capture offset rewind in Kafka f…

### PR DESCRIPTION
**Pull request by lluwm**
…or user stores

In a Kafka problem hit previous, if durability related configs are not set correctly, Kafka could lose data and reset end-offset of the impacted partitions to smaller offsets; new writes on the impacted partitions will have old(small) offsets than the checkpoint offsets in Venice, and thus Venice drops and loses new writes. (Another side effect is that Venice servers could re-consume the entire partition from the beginning). We want to detect this end-offset drop/rewind event and report when it happens. One way to detect it from the Venice servers's point is to compare its local persisted offset for that partition with the end-offset in Kafka and report when the persisted one is greater. The comparison is only done at the time when server subscribes to that partition.

This rb also introduces a new version level metrics to report the offset rewind (backup, current, future, and total) for each user store and it records the occurrence number for the rewind events across all partitions for the store versions. A unit test is also added to verity the behaviour of the metrics is as expected.

Note that this rb only targets store's version topics and the realtime topic is handled by the existing logic e.g. LeaderFollowerStoreIngestionTask.checkAndHandleUpstreamOffsetRewind 

## How was this PR tested?

Internal CI test.
Added new unit tests.

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

**Comment by xunyin8** [comment]
Typo in the RB description for "... when server subsscribes ..."

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 1]
```diff
@@ -1881,6 +1882,8 @@ protected void processCommonConsumerAction(
           startConsumingAsLeader(newPartitionConsumptionState);
         } else {
           updateLeaderTopicOnFollower(newPartitionConsumptionState);
+          reportUserStoreOffsetRewindMetrics(newPartitionConsumptionState);
```
Comment (xunyin8) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:1885]: Looks like we are only doing this rewind check for version topics and not RT topics. Is this the complete scope of the ticket/fix? If so we should probably make it clear in the description or commit message.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 2]
```diff
@@ -2000,6 +2003,58 @@ protected void processCommonConsumerAction(
     }
   }
 
+  /**
+   * This function checks, for a regular user store, if its persisted offset is greater than the end offset of its
+   * corresponding Kafka topic (indicating an offset rewind event, thus a potential data loss in Kafka) and increases
+   * related sensor counter values.
+   */
+  private void reportUserStoreOffsetRewindMetrics(PartitionConsumptionState pcs) {
+    if (!isUserSystemStore()) {
```
Comment (xunyin8) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2012]: Why are we excluding the user system stores? Is there a technical challenge? User system stores are still important and should be treated as first class citizens.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 3]
```diff
@@ -2000,6 +2003,58 @@ protected void processCommonConsumerAction(
...
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    if (offset == OffsetRecord.LOWEST_OFFSET) {
+      return;
+    }
+    // Proceed if persisted OffsetRecord exists and has meaningful content.
+    long endOffset = getKafkaTopicPartitionEndOffSet(localKafkaServer, kafkaVersionTopic, pcs.getPartition());
+    if (endOffset != StatsErrorCode.LAG_MEASUREMENT_FAILURE.code && offset > endOffset) {
+      // report offset rewind.
+      LOGGER.warn(
+          "Offset rewind for version topic: {}, persisted record offset: {}, Kafka topic partition end-offset: {}",
```
Comment (xunyin8) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2025]: We should also log the partition number.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 4]
```diff
@@ -2000,6 +2003,58 @@ protected void processCommonConsumerAction(
...
+      versionedIngestionStats.recordIngestionOffsetRewind(storeName, versionNumber);
+    }
+  }
+
+  /**
+   * @return the end offset in kafka for the topic partition in SIT.
+   */
+  protected long getKafkaTopicPartitionEndOffSet(String kafkaUrl, String kafkaVersionTopic, int partition) {
+    long offsetFromConsumer = getPartitionLatestOffset(kafkaUrl, kafkaVersionTopic, partition);
+    if (offsetFromConsumer >= 0) {
```
Comment (xunyin8) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2038]: I saw you comparing against `OffsetRecord.LOWEST_OFFSET` earlier. Maybe we should try to use the same comparison if the semantic of the check is the same?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 1 above)
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:1885]: Thanks Xun for the review. You are right, this change is targeting the VT topic only. For offset rewind issues in RT topic, I assume it has been taken care by LeaderFollowerStoreIngestionTask.checkAndHandleUpstreamOffsetRewind. Yeah, I need to update the description and make it more clear.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 2 above)
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2012]: This is a good catch! It should be applicable to all stores as you reasoned.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 3 above)
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2025]: Good point.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 4 above)
Comment (xunyin8) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2035]: I see, and there's no constant from Kafka side for the end offset of an empty topic?

Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2035]: No, probably not. The normal way to identify an empty Kafka topic would be to check if its beginning offset == end offset.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java [hunk 5]
```diff
@@ -148,6 +151,9 @@ public IngestionStats(VeniceServerConfig serverConfig) {
             + stalePartitionsWithoutIngestionTaskCount.getClass().getSimpleName(),
         stalePartitionsWithoutIngestionTaskCount);
 
+    ingestionOffsetRewindSensor = localMetricRepository.sensor(INGESTION_OFFSET_REWIND_COUNT);
+    ingestionOffsetRewindSensor.add(INGESTION_OFFSET_REWIND_COUNT, ingestionOffsetRewindCount);
```
[the remark this answers is not in the document]
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java:155]: Good point and I agree. How about renaming it to VERSION_TOPIC_INGESTION_OFFSET_REWIND_COUNT? Or maybe you have a better one in mind.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 6]
```diff
@@ -2000,6 +2003,55 @@ protected void processCommonConsumerAction(
...
+   * corresponding Kafka topic (indicating an offset rewind event, thus a potential data loss in Kafka) and increases
+   * related sensor counter values.
+   */
+  private void reportStoreVersionTopicOffsetRewindMetrics(PartitionConsumptionState pcs) {
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    if (offset == OffsetRecord.LOWEST_OFFSET) {
+      return;
+    }
+    // Proceed if persisted OffsetRecord exists and has meaningful content.
+    long endOffset = getKafkaTopicPartitionEndOffSet(localKafkaServer, kafkaVersionTopic, pcs.getPartition());
```
Comment (sixpluszero) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2017]: I have a minor performance concern: will this affect metrics reporting performance when we hit Kafka issue such as topic not exist or slowness error? 

(I remember we have encountered slow metrics retrieval and it turned out that it is kafka offset retrieval slowness..but maybe I could be wrong here..)

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (xunyin8) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java]: I think that's a good name, very clear.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 6 above)
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2017]: Jialin, thank you for your review and raising this for discussion. To better understand your concerns, do you worry about this function could end up reporting too many metics? or do you worry that getKafkaTopicPartitionEndOffSet could be very slow, thus making the latency of calling this method very high?

For point #1, this metric check happens when server subscribes to a new topic partition and only reports when a rewind is detected, which in normal case, shouldn't be very often.

For point #2, getKafkaTopicPartitionEndOffSet first tries to get the end offset from the `KafkaConsumerService` and thus `TopicPartitionsOffsetsTracker` who knows the offset kind of free when consuming a Kafka record. If it is not available there, then it asks for `CachedKafkaMetadataGetter` who queries Kafka servers, but also caches Kafka responses for 60 seconds for reuse.

Does it make sense, or maybe you referred to something else?

Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2017]: Thanks for this detailed replies, I think I was talking about 2, and seems like this is covered well. Thank you! Please feel free to mark it as resolved.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java [hunk 7]
```diff
@@ -53,6 +53,7 @@ public class IngestionStats {
   protected static final String OFFSET_REGRESSION_DCR_ERROR = "offset_regression_dcr_error";
   protected static final String TOMBSTONE_CREATION_DCR = "tombstone_creation_dcr";
   protected static final String READY_TO_SERVE_WITH_RT_LAG_METRIC_NAME = "ready_to_serve_with_rt_lag";
+  public static final String VERSION_TOPIC_INGESTION_OFFSET_REWIND_COUNT = "vt_ingestion_task_offset_rewind_count";
```
Comment (huangminchn) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java:56]: Is this the metric name? I don't think we need to put ingestion key word in it. This metric is used to alert Kafka data loss. Can we rename it to "vt_end_offset_rewind_count" or "version_topic_end_offset_rewind_count"?

**Review by huangminchn — CHANGES_REQUESTED** [review]
Thanks a lot Lei! We need to correct this problem statement a bit:

> one issue we have seen in the past is that Kafka partition on a consumer can lose its offset and start re-consuming the entire partition from the beginning.

To be more accurate, could you please update it to:
If durability related configs are not set correctly, Kafka could lose data and reset end-offset of the impacted partitions to smaller offsets; new writes on the impacted partitions will have old(small) offsets than the checkpoint offsets in Venice, and thus Venice drops and loses new writes. (Another side effect is that Venice servers could re-consume the entire partition from the beginning).

What do you think?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java (hunk 7 above)
Comment (lluwm) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java:56]: Thanks Min for the review. Sounds reasonable.

**Review by huangminchn — APPROVED** [review]

**Pull request merged**